docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,158 @@
+  to obtain required secrets.
+
+* Transmission of the authentication token requires one or more shared file systems
+  between the service and `security-secretstore-setup`.
+  In the Docker implementation,
+  this shared file system is implemented by host-based local storage.
+  Similar functionality is near-impossible in a Kubernetes environment
+  that supports only RWO volumes:
+  the volume containing the secret can be mounted to only a single pod at a time.
+  It is simply easier to run without security enabled in Kubernetes.
+
+* The current approach cannot support distributed services
+  without an underlying distributed file system to distribute tokens,
+  such as GlusterFS, running across the participating nodes.
+
+## Decision
+
+EdgeX will create a new service, `security-spiffe-token-provider`.
+This service will be a mutual-auth TLS service that exchanges
+a [SPIFFE](https://spiffe.io/) X.509 SVID for a secret store token.
+
+The `go-mod-secrets` module will be modified enable a new mode
+whereby a secret store token is obtained by:
+
+1. Obtaining an X.509 SVID by contacting a local SPIFFE agent's
+   workload API on a local Unix domain socket.
+2. Connecting to the `security-spiffe-token-provider` service
+   and using the X.509 SVID and request a secret store token.
+
+The SPIFFE authentication mode will be the default authentication
+mode for non-core EdgeX services and for remote EdgeX services.
+
+### Technical Architecture
+
+![SPIFFE Architecture and Workflow](0019-spiffe-architecture.jpg)
+
+The work flow is as follows:
+
+1. The SPIFFE server is started.
+1. The server creates an initial set of keys.
+1. The trust bundle (certificate authority) data is exported from the SPIFFE server
+   and workload entries are loaded into the SPIFFE server database.
+1. A join token for the SPIFFE agent is created.
+1. The SPIFFE agent is started with the join token to add it to the cluster.
+1. Vault is started and `security-secret-store-setup`
+   initializes it and creates an admin token for `security-spiffe-token-provider` to use.
+1. The `security-spiffe-token-provider` service is started.
+   It obtains an SVID from the SIFFE agent and uses it as a TLS server certificate.
+1. An EdgeX microservice starts and obtains another SVID from the SPIFFE agent
+   and uses it as a TLS client certificate to contact the
+   `security-spiffe-token-provider` service.
+   The EdgeX microservice uses the trust bundle as a server CA
+   to verify the TLS certificate of the remote service.
+1. `security-spiffe-token-provider` verifies the SVID using the trust bundle as client CA
+   to verify the client,
+   extracts the service key,
+   and issues an appropriate Vault service token.
+1. The EdgeX microservice accesses Vault as usual.
+
+## Consequences
+
+This proposal will require addition of several new, optional, EdgeX microservices:
+
+* `security-spiffe-token-provider`, running on the main node
```

Comment:
It could... but you have to choose:

(1) SPIFFE support becomes a required feature of turning on security instead of an optional add-on.

(2) There needs to be a runtime feature enable (disable?) to control whether or not the utility contacts the SPIFFE workload endpoint to get an X.509 SVID.  If we go this route, then we need to regress three EdgeX configurations: security disabled, security enabled without SPIFFE, and securing enable with SPIFFE.

The way I have it spec'ed now one doesn't need to run any of these optional services (save modifications to the bootstrapper) if one doesn't have delayed start services.